### PR TITLE
Implement IConvertible on PyObject

### LINKED ISSUE #1755: Python int type is translated to pyInt when passed as a .net Dictionary value only in 3.x version
### Environment

-   Pythonnet version: 3.0.0a2
-   Python version: 3.7, 3.8, 3.9, 3.10
-   Operating System: Windows
-   .NET Runtime: 4.7.2 framework

### Details

After upgrading to pre release 3.0.0a2 version, the python int type is getting translated to PyInt instead of System.int32 inside the .NET DLL, and hence getting a unable to cast exception. this is working fine in 2.x version.

DLL method: 
```csharp
public void DictMethod(Dictionary<string, object> paramDict)
  {
            Console.WriteLine("Inside");
            Console.WriteLine(paramDict["index"].GetType().ToString());
            var index = Convert.ToInt32((paramDict["index"]));
            
        }
```
Python script: 

```python
   clr.AddReference("System.Collections")
from System.Collections.Generic import Dictionary
from System import String
from System import Object
clr.AddReference("PyTest")
from PyTest import PyClass

class Utils(object):

    @staticmethod
    def prepare_dict(**kwargs):

        dict_net = Dictionary[String,Object]()
        for key_value in kwargs.keys():
            dict_net.Add(key_value,kwargs[key_value])
        return dict_net

parameters = {}
parameters['index'] = 1
net_dict = Utils.prepare_dict(**parameters)
obj = PyClass()
obj.DictMethod(net_dict)
```

Output in 3.x version:
```
C:\Users\Athul\Desktop>c:\Python37\python.exe pynettest.py
Inside
Python.Runtime.PyInt
Traceback (most recent call last):
  File "pynettest.py", line 50, in <module>
    obj.DictMethod(net_dict)
System.InvalidCastException: Unable to cast object of type 'Python.Runtime.PyInt' to type 'System.IConvertible'.
   at System.Convert.ToInt32(Object value)
   at PyTest.PyClass.DictMethod(Dictionary`2 paramDict) in C:\Users\Athul\source\repos\PyTest\PyClass.cs:line 46
```

Output in 2.x version : 

```
C:\Users\Athul\Desktop>c:\Python37\python.exe pynettest.py
Inside
System.Int32

C:\Users\Athul\Desktop>
```

### PR DESCRIPTION
**Pull request by filmor**
### What does this implement/fix? Explain your changes.

Implements the [IConvertible](https://docs.microsoft.com/en-us/dotnet/api/system.iconvertible) interface for `PyObject` instances.

### Does this close any currently open issues?

Should fix #1755.

### Any other comments?

...

### Checklist

Check all those that are applicable and complete.

-   [x] Make sure to include one or more tests for your change
-   [X] If an enhancement PR, please create docs and at best an example
-   [x] Ensure you have signed the [.NET Foundation CLA](https://cla.dotnetfoundation.org/pythonnet/pythonnet)
-   [x] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [ ] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

### CODE REVIEW

**Comment by lostmsu** [comment]
.NET team now avoids `IConvertible`: https://github.com/dotnet/runtime/issues/936#issuecomment-663743656

**Comment by lostmsu** [comment]
See also https://github.com/dotnet/runtime/issues/62022

**Comment by filmor** [comment]
It doesn't look like there is an alternative yet, though, right? And I also don't think that they are actually going to drop the interface, they are just not extending it for new primitive types, which have no representation in Python anyway.

**Comment by lostmsu** [comment]
@filmor the new `Half` type does not implement `IConvertible` though it is nearly identical to `Single` and `Double`. So I guess it is kinda deprecated (not that I completely agree with everything .NET team does).

File: src/runtime/PythonTypes/PyInt.cs [hunk 1]
```diff
@@ -230,5 +230,7 @@ public string ToString(string format, IFormatProvider formatProvider)
             using var _ = Py.GIL();
             return ToBigInteger().ToString(format, formatProvider);
         }
+
+        public override TypeCode GetTypeCode() => TypeCode.Int64;
```
Comment (lostmsu) [inline src/runtime/PythonTypes/PyInt.cs:234]: Generally, this is wrong.

Comment (filmor) [reply src/runtime/PythonTypes/PyInt.cs:234]: The docs are a bit unclear on this function. I used this as an "should almost always work" value, the `InvalidCastException` might still be thrown if the conversion fails. We could drop the `GetTypeCode` overrides entirely and just return `TypeCode.Object` if you'd prefer that.

Comment (lostmsu) [reply src/runtime/PythonTypes/PyInt.cs:234]: I can't say for others, just for this one. Honestly, not sure if anyone ever uses this.

Comment (filmor) [reply src/runtime/PythonTypes/PyInt.cs:234]: I'm not sure I follow. `GetTypeCode` is supposed to be a hint that allows you to write a `switch` like
```csharp
switch (convertible.GetTypeCode()) {
    case TypeCode.Int64:
        return Convert.ToInt64(convertible);
}
```
It doesn't mean that the object *is* a `System.Int64` already.

Comment (lostmsu) [reply src/runtime/PythonTypes/PyInt.cs:234]: You seem to be correct on that. Still not sure `Int64` is valid for this specific case, but float and string probably are. I am fine either way though. I don't think we will get people complaining one way or another any time soon.

**Pull request merged**